### ISSUE TITLE
Carbonado Oven Becomes Item Duplication Machine Instead of Oven...

### ISSUE THREAD

**Issue by mrcoffee1026** [TheBusyBiscuit/Slimefun4#512]
## Description
Oven from Enhanced+ run same speed as Vanilla/unmodified oven, although efficiency is increased as levels increase.  At some random interval, oven will stop producing cooked output and instead begin dropping an (infinite?) amount of whatever was put in to be cooked along with what was put in to be used as fuel.  (Use of hoppers to duplicate this issue is optional but does seem to speed up the appearance of the effect).

Issue originally appeared in SF v4.1.10, the v4.1.11's release notes elluded to the possibility this issue had been dealt with, but it clearly has not been.

## Server Log
Nothing relevant, there are no errors, things are doing what they were programmed to do.

## Evidence / Proof
https://imgur.com/a/VuQ3h

## Environment
CraftBukkit version
git-Spigot-3d850ec-19ed80c (MC: 1.12.2) (Implementing version 1.12.2-R0.1-SNAPSHOT)
CS-CoreLib v1.5.16
Slimefun v4.1.11
ExoticGarden v1.6.3

**Comment by Poslovitch** [comment]
Duplicate of https://github.com/TheBusyBiscuit/Slimefun4/issues/411.

I'm starting to think it may be caused by the 10x speed bonus.

Nevermind, disable them on your server to avoid duplication, we didn't find any fix for it for 2 months.

**Comment by mrcoffee1026** [comment]
Caused by.. speed that doesn't get applied anymore?  Also all of the "faster" ovens are having the same issue, it would seem that it would be necessary to disable ALL slimefun ovens if that is the "solution" to this.

**Comment by Poslovitch** [comment]
This is currently the only solution we can provide to you. As far as we know, this is a Minecraft or Spigot bug. And up to now we havent found a better workaround than disabling the Enhanced furnaces.

**Comment by mrcoffee1026** [comment]
A Spigot/Minecraft bug... that only effects slimefun v4.1.10 and up but didn't effect 4.1.9?  That seems rather unlikely... I think the changes made in 4.1.10 should be more closely examined.

**Comment by mrcoffee1026** [comment]
Disabling them isn't really a solution either since already-made ovens... still have the issue and can't be 'unmade'.

**Comment by Poslovitch** [comment]
Are you sure it comes from 4.1.10 exactly? Not due to a MC update? I might investigate it more closely

**Comment by mrcoffee1026** [comment]
I am absolutely certain the duplication issue does not exist in 4.1.9.  I have rolled back to this version to avoid the problem.  The speed issue is not the same, if you want to blame that on MC/Spigot... well, I can't deter you from that course of action and can't be assured it is not related in some way to MC since speed didn't fail till MC 1.11+, but the duplication issue is a direct result of SF 4.1.10.

**Comment by Poslovitch** [comment]
Thanks for your answer. I'll do a complete check to find where the bug may have come from.

**Comment by mrcoffee1026** [comment]
you're welcome, i'm anxious to know what you discover...

**Comment by mrcoffee1026** [comment]
Verified this glitch is ubiquitous throughout the "enhanced" furnaces (not just reinforced/carbonado).

**Comment by SpiritAmet** [comment]
is there any collaborative plans to have this issue resolved by chance?

**Comment by Poslovitch** [comment]
Currently, nope, because all the data gathered by tests is beyond understanding : see https://github.com/TheBusyBiscuit/Slimefun4/issues/453#issuecomment-347972873

**Comment by SpiritAmet** [comment]
Thanks for the info Poslovitch, so at this point can we get a list of all duplicating items, that we may be able to disable them please and thank you.  As well as an explanation on how to disable (i just want to make sure I am following things correctly instead of walking blindly in the dark.

**Comment by Poslovitch** [comment]
In Items.yml

for each ENHANCED_FURNACE, set "enabled" to false

**Comment by John000708** [comment]
Can someone who had the duplication issue download the last Slimefun version from https://thebusybiscuit.github.io/builds/TheBusyBiscuit/Slimefun4/master/ and see whether they still have the problem?

**Comment by runelynx** [comment]
@John000708 Will do that now. I'll post back here if we see it again.

**Comment by mrcoffee1026** [comment]
I retested carbonado which usually begins vomiting supplies nearly instantly in a hopper situation... and not only did it function without doing so, it did so at speed.  Using hoppers still completely disables 'luck' but I'm happy to see at least the oven itself working at speed again without item duplication.  Will take awhile longer to retest all of them, the lower level furnaces seemed to operate a long while without incident and then began dropping supplies later, so if they still make a mess they'll take longer to figure out, but you've probably nailed it since the carbonado quit doing it from what I can see.  Awesome!

**Comment by mrcoffee1026** [comment]
Duplication / crazy dropping of fuel and item to be smelted still occurs when smelting armor items (example disenchanted 12-c gold boots).

**Comment by John000708** [comment]
@mrcoffee1026 can you give me a list of your plugins? Since you can smelt armor pieces I assume that you're using a plugin with which you can customize recipes. I guess ProRecipes which would be related with #278

**Comment by mrcoffee1026** [comment]
Iron/gold Armor and tools have been smeltable since 1.10 -- completely Vanilla, this is not a plugin feature.  No plugin provides that functionality that I am aware of. @John000708

**Comment by mrcoffee1026** [comment]
also duplicates when cooking fish... so in fact the only items that don't duplicate like crazy so far is iron ore, gold ore, and sand.

**Comment by John000708** [comment]
@mrcoffee1026 do you have a server I can join to test this issue? I cannot reproduce it locally.

**Comment by mrcoffee1026** [comment]
@John00070 sure?  communitybrew.net

**Comment by mrcoffee1026** [comment]
Oh um... it appears that this glitch is gone if you don't use hoppers?  But never fails to occur if there are hoppers involved... this is the most basic hopper set up which will result in the described behavior:
https://imgur.com/YVblG41
@John000708

**Comment by Poslovitch** [comment]
Sounds like the hoppers are our guilty guy! Let's ask Mojang to remove them mwahaha.
Afaik, though, I've never seen any hopper-related processing in Furnaces listeners

**Comment by mrcoffee1026** [comment]
Well they certainly do alter the way these furnaces operate... as I've mentioned before placing a hopper on output effectively disables the "luck" aspect of these ovens as well, so there is something they're not doing properly with hoppers and haven't been for some time... it's just that since adding a hopper to the oven makes it an evil duplication engine (and mess creator), it does still need to be dealt with at some point, there's almost no case in a which someone would have a carbonado oven and NOT use hoppers for input/output.

**Comment by John000708** [comment]
@mrcoffee1026 I'm currently on the server. If you could come on for a bit I'd be happy to check it out

**Comment by mrcoffee1026** [comment]
on now...  you can hit me on our discord if we continue to have issues missing each other: https://discord.gg/QPVHMJe

**Comment by mrcoffee1026** [comment]
fascinating: if you place a minecart hopper instead of a regular one for the output, it also does not duplicate.

**Comment by mrcoffee1026** [comment]
You seem to have fixed it?  Possibly I just needed to replace any ovens that had been placed prior to your fix... I'll keep poking at it every so often, but I've been unable to cause the issue again since simply breaking and replacing the ovens, protections or not.  I thought for sure I'd placed a new oven and hopper and saw the problem since the updated download but... I can't swear to that.  Unless someone else runs into it or if I see it again I think you may have nailed it.  It's unfortunate hoppers disable the luck factor, still... but there it seems that the oven relies on its internal count to provide that enhancement for whatever reason so there may be nothing that can be done about it.

**Comment by mrcoffee1026** [comment]
This issue does recur after every restart.  I find I have to make a griefprevention claim around a furnace then break and replace every other furnace so that they cease this behavior.  After the replace, they seem fine.

**Comment by mrcoffee1026** [comment]
@John000708 After our discussion and tests on things and you're asking about my LWC version... i tried looking for both the version I ran and finding a more updated one.  I couldn't find the one that I ran, nor can I find documentation on a 4.6.0-Snapshot build anywhere, so gave up and went with the "unofficial" entity locking version since they posted an update in December.  This didn't have any immediate effect, but after having to restart a number of times to make changes to that new config, I haven't had to break and replace an oven since.  I've not been able to duplicate the issue and am not sure why I had to replace a bunch of ovens at all since the problem appears to be resolved.  If it creeps back up again I'll try to do some more testing but I think you got it, and for some reason the fix had a delayed onset for me.

**Comment by mrcoffee1026** [comment]
sorry another revision... this does continue to recur with ARMOR AND TOOL items until oven has been broken and replaced.

**Comment by mrcoffee1026** [comment]
This is absolutely still happening.  Methods of duplication: let players make them.  Eventually they'll begin doing this on their own whether hoppers have been used or not.  Everything else it seems to be a game of probability... the likelihood of becoming a messy duplication machine increases when there is a hopper on the output, when the server has just been restarted, and when armor or tools are being smelted.  You can decrease the probability by breaking the oven WHILE it is duplicating items... and re-placing it.  You can further decrease the probability of duplication by not using hoppers (but there is no way of preventing everyone from using hoppers on this device since it still qualifies as a vanilla furnace).  You can even further decrease the probability by never smelting armor or tools.  Nothing eliminates the issue, it eventually begins occurring whether or not any of those things are going on... on furnaces without hoppers and during the smelting of something like iron ore.  There is no correlation that I can find as to time of the last restart, that is simply how it is.  The issue has not been resolved and will definitely require further investigation.

**Comment by mrcoffee1026** [comment]
Absolutely still happening to all slimefun furnaces whether or not hoppers are used.  User had placed a new one, and within 2 hours it was vomiting duplicated fuel and iron ore onto the floor, which was not an item that usually resulted in this issue... and in a non-hopper set up which also does not usually result in this issue.  I've found that using a minecart hopper on the output slot seems to "fix" the issue... however, this is not a reliable solution since users are unlikely to take this approach, and may well want the duplication issue... in which case they certainly would not.  Also any use of hoppers on the output slot automatically destroys the oven's ability to apply the luck factor to iron and gold which is one of the major benefits to using this oven to begin with.  Typically breaking and replacing the oven causes it to function normally... but only for some random length of time before it starts all over again.  Bug is as alive now as it ever was as of dev build 4-19.

**Comment by Deathinflames** [comment]
We've also been struggling with this issue, and while disabling the furnaces altogether is a workaround we've been trying to avoid this as much as possible due to a lot of other things in Slimefun that we've had to disable for various other reasons.

While we've also been unable to find a way to make this issue occur intentionally so that it can be fixed we strongly believe that the root of the issue is due to how Slimefun saves data via flatfile. It has been our experience in the years of hosting a minecraft server that large plugins which save large amounts of data to flatfile can at times have inconsistencies which result in either strange occurrences and in more severe cases even data corruption. 

We've discovered some unused code(or as far as we can tell it is unused), in the CS-CoreLib repository  that would allow compatibility for MySQL saving for Slimefun and for the rest of the plugins made by TheBusyBiscuit. We strongly believe that most if not all of these kinds of issues would be resolved if that functionality was finished. It's a no brainer that for large amounts of data that needs to be saved - MySQL is the best option. Unfortunately our skill with Java is sub par to be able to contribute with making this functionality a reality, we thought we'd just throw this out there anyways for consideration and to possibly give a different perspective on what might actually be happening and why it is so random.

**Comment by mrcoffee1026** [comment]
Looks like the trick of using a minecart hopper on the output doesn't always work either.  Has failed on brand new ovens 2 hours after place (no restart) while smelting cobble to stone (no luck on that either, so.......), similar vomiting disasters post restart with iron ore, cobble again, various fuels in and out of a hopper situation.  It's not very consistent, sometimes I go days not restarting without any issues and then... all at once there's a bunch.  Anyway I have to once again disable all sf furnaces till this is resolved.

**Comment by mrcoffee1026** [comment]
I went to the extreme step of disabling these and installing a different plugin for furnaces.  I then fine tuned each level of those furnaces to match speed from the slimefun furnaces, requiring me to test many smelts... and they would randomly vomit duplicated input all over the ground right up to the very end of doing this... and I had to do this 2-3 times as they continued to release new builds of that plugin, a few in regards to feedback from myself as I found an identical sitionion with those: speed and output "luck" is removed when hoppers are used.    I don't know how the innards of the plugin differ nor what the dev did to fix it, but she was able to provide a remedy for that issue almost immediately which lets me know at least: it's a fixable issue and is not "Spigot's fault" which will just go away on its own someday.  Besides that repairing what's wrong with these may simply be beyond me right now... since Slimefun releasing for 1.11, these furnaces - even when "working" dropped their luck by about 1/3 and only apply it to ore blocks now.  By nerfing these machines so completely, I lack the motivation to reverse engineer what's going wrong with them if that is truly working as intended.  Unless our favorite dev steps in steps in and says "yes" they should be working on all blocks and outputting nearly 3X as much... with or without hoppers... if so I'll revoke that statement and start tinkering with the why they sometimes explode items everywhere and why hoppers kill them.  Till then... I've managed a workaround wherein I use furnaces from another plugin which players "buy" using the same ingredients the slimefun furnaces required to craft as currency.
